text/0007-mfa-rollout.md

```diff
@@ -0,0 +1,150 @@
+```
+
+Example message for critical gem commands (`gem push/yank`, and `gem owner -a/r`) for the 100 most downloaded gem accounts
+```
+[WARNING] For protection of your account and gems, we encourage you to set up multi-factor authentication at https://rubygems.org/multifactor_auth/new. Your account will be required to have MFA enabled in the future.
+```
+
+## Phase 3: Enforcing MFA on Most Downloaded Gems (est. July/Aug 2022)
+
+At a future date, yet to be determined, we will require MFA to be enabled in order to interact with Rubygems.org. Enforcement will include the following changes.
+
+On the UI, after login, the setup MFA page will be displayed for users to setup MFA. They are able to back out, but the user will be directed to set up MFA whenever they choose to visit a page that requires authentication. A banner will also appear describing the intent. 
+
+Example message
+```
+For protection of your account and your gems, you are required to set up multi-factor authentication.
+```
+
+On the CLI, on `gem signin`, `gem push/yank`, and `gem owner -a/r`, an error message will display to require users to setup MFA on rubygems.org and will block the request from succeeding. 
+
+Example message
+```
+For protection of your account and your gems, you are required to set up multi-factor authentication at https://rubygems.org/multifactor_auth/new
+```
+
+## Phase 4: Rolling out to the General Public (Q4 2022 and beyond)
+
+Once the rollout of the top 100 gems is successfully rolled out, a similar rollout will be done to another cohort of gems until all accounts are required to have MFA enabled. 
+
+# Reference-level explanation
+
+## API Key and Gem Publishing Improvements
+The API key model will need to be changed in order to implement these improvements. For new features like adding gem scope and expiry date on keys, columns will be added for expiry and gem scope and additional options will need to be added to the CLI and the UI. Setting an expiry on existing keys can be considered a breaking change.
+
+Gem publishing will be improved by allowing gem owners to push mulitple gems at once, so that they only need to enter their OTP code one time. A prototype for pushing multiple gems has been briefly explored here:  https://github.com/Shopify/rubygems/pull/2
+
+## Policy Rollout
+A user account will be required to have MFA enabled if they own a gem that exceeds a download threshold of 175 million. This number was selected as it is around the total downloads of the 100th most downloaded gem ([sinatra](https://rubygems.org/gems/sinatra)). There will be two methods introduced in the Rubygems model. One would determine whether the gem owners are encouraged to setup MFA (eg. `mfa_recommended?`) and display relevant warnings on the UI/CLI. The other would determine if gem owners are required to setup MFA (eg. `mfa_required?`) and start blocking sensitive behaviours.
```

Comment:
> We can also ping the authors not using MFA in advance when 90% (or any reasonable part) of threshold is reached. (from https://github.com/rubygems/rfcs/pull/36#issuecomment-1024894144)

I don't have much context around the rate of downloads but would it make sense to warn/ping users ASAP at around ~175 million, and enforce at ~180 million in July/Aug (which hopefully encompasses `sinatra` at that point). This would prevent owners on the border to be surprised once they reach the required threshold.

Reply:
I like the idea of introducing a signal earlier before gem owners have reached the 175 million threshold. If we do implement this, I'd recommend warning them sooner than ~170 million downloads since by that point they'll be ~97% away from reaching the ~175 million threshold. 

The guidance of ~90% away from threshold seems reasonable. That would translate to warning ~157 million downloads.

Reply:
👍 That makes sense, I updated the thresholds to 180 million for required since `sinatra` (100th most downloaded gem) has reached that mark and 165 million for recommended. 165 million is roughly 90% of 180 million.